src/hal/src/external_memory/mod.rs

```diff
@@ -0,0 +1,263 @@
+//! Structures related to the import external memory functionality
+
+mod errors;
+pub use errors::*;
+
+#[cfg(any(unix))]
+#[cfg_attr(feature = "unstable", doc(cfg(unix)))]
+mod fd;
+#[cfg(any(unix))]
+#[cfg_attr(feature = "unstable", doc(cfg(unix)))]
+pub use fd::*;
+
+#[cfg(any(windows))]
+#[cfg_attr(feature = "unstable", doc(cfg(windows)))]
+mod handle;
+#[cfg(any(windows))]
+#[cfg_attr(feature = "unstable", doc(cfg(windows)))]
+pub use handle::*;
+
+mod ptr;
+pub use ptr::*;
+
+/// Buffer or image
+#[derive(Debug)]
+pub enum BufferOrImage<'a, B: crate::Backend> {
+    /// Buffer
+    Buffer(&'a B::Buffer),
+    /// Image
+    Image(&'a B::Image),
+}
+
+/// External buffer properties
+#[derive(Debug, PartialEq)]
+pub struct ExternalBufferProperties {
+    usage: crate::buffer::Usage,
+    sparse: crate::memory::SparseFlags,
+    external_memory_properties: ExternalMemoryProperties,
+}
+impl ExternalBufferProperties {
+    /// Constructor
+    pub fn new(
+        usage: crate::buffer::Usage,
+        sparse: crate::memory::SparseFlags,
+        external_memory_properties: ExternalMemoryProperties,
+    ) -> Self {
+        Self {
+            usage,
+            sparse,
+            external_memory_properties,
+        }
+    }
+    /// Is the queried configuration exportable
+    pub fn get_queried_buffer_usage(&self) -> crate::buffer::Usage {
+        self.usage
+    }
+    /// Is the queried configuration importable
+    pub fn get_queried_buffer_sparse(&self) -> crate::memory::SparseFlags {
+        self.sparse
+    }
+    /// Get external memory properties
+    pub fn get_external_memory_properties(&self) -> &ExternalMemoryProperties {
+        &self.external_memory_properties
+    }
+}
+impl AsRef<ExternalMemoryProperties> for ExternalBufferProperties {
+    fn as_ref(&self) -> &ExternalMemoryProperties {
+        &self.external_memory_properties
+    }
+}
+impl std::ops::Deref for ExternalBufferProperties {
+    type Target = ExternalMemoryProperties;
+    fn deref(&self) -> &Self::Target {
+        &self.external_memory_properties
+    }
+}
+
+/// External memory properties
+#[derive(Debug, PartialEq)]
+pub struct ExternalMemoryProperties {
+    exportable: bool,
+    importable: bool,
+    dedicated_allocation: bool,
+    memory_type: ExternalMemoryType,
+    compatible_memory_types: ExternalMemoryTypeFlags,
+    export_from_imported_memory_types: ExternalMemoryTypeFlags,
+}
+impl ExternalMemoryProperties {
+    /// Constructor
+    pub fn new(
+        exportable: bool,
+        importable: bool,
+        dedicated_allocation: bool,
```

Comment:
As an example of trying to reduce the API surface here, would it make sense to just limit the API to dedicated allocations only? This should be simpler, and yet I think it would cover all the cases at least Gecko cares about.

Reply:
Assuming that every external memory would be a dedicated allocation would simplify a lot the API. Since they could not be suballocated, makes no sense to keep separated the functions for the creation of external buffer and external memory, so that the two functions that do this could be merged toghether while reducing the chance of wrong api utilization (so improving safety).

Reply:
Wonderful. I hope you don't mind reducing the scope a bit here :)

Reply:
No problem, i have already merged the functions and i'm working to fix all the other problems. I will push the new code soon

Reply:
The new code has been pushed. I have also modified the test application to match the new api. Since the new api is more concentrated and makes more integrated checks, the test application code has been reduced to 1/3 of the previous version, so this means the effort to use this feature has been greatly reduced.